pkg/apis/datasource/v0alpha1/types.go

```diff
@@ -0,0 +1,113 @@
+package v0alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/grafana/grafana/pkg/apis"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+)
+
+const (
+	GROUP   = "*.datasource.grafana.app"
+	VERSION = "v0alpha1"
+)
+
+var GenericConnectionResourceInfo = apis.NewResourceInfo(GROUP, VERSION,
+	"connections", "connection", "DataSourceConnection",
+	func() runtime.Object { return &DataSourceConnection{} },
+	func() runtime.Object { return &DataSourceConnectionList{} },
+)
+
+var GenericConfigResourceInfo = apis.NewResourceInfo(GROUP, VERSION,
+	"config", "config", "DataSourceConfig",
+	func() runtime.Object { return &DataSourceConfig{} },
+	func() runtime.Object { return &DataSourceConfigList{} },
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DataSourceConnection struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The display name
+	Title string `json:"title"`
+
+	// Optional description for the data source (does not exist yet)
+	Description string `json:"description,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DataSourceConnectionList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []DataSourceConnection `json:"items,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DataSourceConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The generic datasource config
+	Spec ConfigSpec `json:"spec"`
+
+	// Generic secure keys -- this will accept strings as input,
+	// but the results will become keys into a secret store
+	Secure SecureSpec `json:"secure"`
+}
+
+type ConfigSpec struct {
+	Name     string `json:"name"`
+	Access   string `json:"access,omitempty"` // proxy??
+	URL      string `json:"url,omitempty"`
+	ReadOnly bool   `json:"readOnly,omitempty"`
+
+	User            string `json:"user"`
+	Database        string `json:"database"`
+	BasicAuth       bool   `json:"basicAuth,omitempty"`
+	BasicAuthUser   string `json:"basicAuthUser"`
+	WithCredentials bool   `json:"withCredentials,omitempty"` // ???
+
+	// The public generic config data
+	JsonData *simplejson.Json `json:"jsonData"`
```

Comment:
the go type for this one seems specific to grafana internals. is there a clean way to avoid this?

Reply:
good question -- this is essentially a `map[string]any` with helpers.  naively adding map[string]any requires custom logic for deepCopy and openapi

I think we need *something* in apis exposed that supports this behavior -- but not sure it needs to be simplejson specifically